Mathlib/FieldTheory/Galois/InfiniteDimensional.lean

```diff
@@ -0,0 +1,212 @@
+
+instance (L₁ L₂ : IntermediateField k K) [FiniteDimensional k L₂] :
+    FiniteDimensional k ↑(L₁ ⊓ L₂) :=
+  .of_injective (IntermediateField.inclusion inf_le_right).toLinearMap
+    (IntermediateField.inclusion inf_le_right).injective
+
+instance (L₁ L₂ : IntermediateField k K) [Algebra.IsSeparable k L₁] :
+    Algebra.IsSeparable k ↑(L₁ ⊓ L₂) :=
+  .of_algHom _ _ (IntermediateField.inclusion inf_le_left)
+
+instance (L₁ L₂ : IntermediateField k K) [Algebra.IsSeparable k L₂] :
+    Algebra.IsSeparable k ↑(L₁ ⊓ L₂) :=
+  .of_algHom _ _ (IntermediateField.inclusion inf_le_right)
+
+instance (L₁ L₂ : IntermediateField k K) [IsGalois k L₁] [IsGalois k L₂] :
+    IsGalois k ↑(L₁ ⊓ L₂) := {}
+
+instance : Sup (FiniteGaloisIntermediateField k K) where
+  sup L₁ L₂ := .mk <| L₁ ⊔ L₂
+
+instance : Inf (FiniteGaloisIntermediateField k K) where
+  inf L₁ L₂ := .mk <| L₁ ⊓ L₂
+
+instance : Lattice (FiniteGaloisIntermediateField k K) :=
+  val_injective.lattice _ (fun _ _ ↦ rfl) (fun _ _ ↦ rfl)
+
+instance : OrderBot (FiniteGaloisIntermediateField k K) where
+  bot := .mk ⊥
+  bot_le _ := bot_le (α := IntermediateField _ _)
+
+@[simp]
+lemma le_iff (L₁ L₂ : FiniteGaloisIntermediateField k K) :
+    L₁ ≤ L₂ ↔ L₁.toIntermediateField ≤ L₂.toIntermediateField :=
+  Iff.rfl
+
+variable (k) in
+/-- The finite Galois intermediate field obtained from the normal closure of adjoining a
+`s : Set K` to `k`. -/
+noncomputable def adjoin [IsGalois k K] (s : Set K) [Finite s] :
+    FiniteGaloisIntermediateField k K := {
+  normalClosure k (IntermediateField.adjoin k (s : Set K)) K with
+  finiteDimensional :=
+    letI : FiniteDimensional k (IntermediateField.adjoin k (s : Set K)) :=
+      IntermediateField.finiteDimensional_adjoin <| fun z _ =>
+        IsAlgebraic.isIntegral (Algebra.IsAlgebraic.isAlgebraic z)
+    normalClosure.is_finiteDimensional k (IntermediateField.adjoin k (s : Set K)) K
+  isGalois := IsGalois.normalClosure k (IntermediateField.adjoin k (s : Set K)) K }
+
+@[simp]
+lemma adjoin_val [IsGalois k K] (s : Set K) [Finite s] :
+    (FiniteGaloisIntermediateField.adjoin k s) =
+    normalClosure k (IntermediateField.adjoin k s) K :=
+  rfl
+
+variable (k) in
+lemma subset_adjoin [IsGalois k K] (s : Set K) [Finite s] :
+    s ⊆ (adjoin k s).toIntermediateField :=
+  (IntermediateField.subset_adjoin k s).trans (IntermediateField.le_normalClosure _)
+
+@[simp, nolint simpNF]
+theorem adjoin_le_iff [IsGalois k K] {s : Set K} [Finite s]
+    {L : FiniteGaloisIntermediateField k K} : adjoin k s ≤ L ↔ s ≤ L.toIntermediateField := by
+  simp only [le_iff, adjoin_val, IntermediateField.normalClosure_le_iff_of_normal,
+    IntermediateField.adjoin_le_iff, Set.le_eq_subset]
+
+@[simp, nolint simpNF]
+theorem adjoin_simple_le_iff [IsGalois k K] {x : K} {L : FiniteGaloisIntermediateField k K} :
+    adjoin k {x} ≤ L ↔ x ∈ L.toIntermediateField := by
+  simp only [adjoin_le_iff, Set.le_eq_subset, Set.singleton_subset_iff, SetLike.mem_coe]
```

Comment:
We should avoid `nolint simpNF`. Either remove `@[simp]` from `le_iff` or state these with `.toIntermediateField` (or rather coerced).

Reply:
What do you mean by remove `@[simp]`? remove these two `@[simp]` or what?

Reply:
Remove the `@[simp]` line from the `FiniteGaloisIntermediateField.le_iff` lemma that you added above. This is why the `simpNF` linter complains: the left hand sides of these two lemmas are not in `simp` normal form, i.e. there is a `simp` lemma somewhere (in this case `FiniteGaloisIntermediateField.le_iff`) that applies to the left hand side of your lemmas and changes it to something syntactically different. This means that these lemmas will probably never trigger in a `simp` call as the LHS will be simplified to something else first.

Reply:
Remove the `@[simp]` line from the `FiniteGaloisIntermediateField.le_iff` lemma that you added above. This is why the `simpNF` linter complains: the left hand sides of these two lemmas are not in `simp` normal form, i.e. there is a `simp` lemma somewhere (in this case `FiniteGaloisIntermediateField.le_iff`) that applies to the left hand side of your lemmas and changes it to something syntactically different. This means that these lemmas will probably never trigger in a `simp` call as the LHS will be simplified to something else first.

Reply:
Thank you, I would prefer to resolve this way.

Reply:
But the second one still have normal form problem (I think it is because it used the first one), how to resolve? Thank you.

Reply:
Ah, probably the second one is solvable by `simp` already, so in this case either remove `@[simp]` from `adjoin_simple_le_iff` or remove `adjoin_simple_le_iff` altogether. I think `adjoin_simple_le_iff` is not needed.

Reply:
I think the best way is to
- keep `le_iff` a `simp` lemma
- remove `@[simp]` from `adjoin_le_iff` (as it is now provable by `simp`)
- remove `adjoin_simple_le_iff`

Reply:
Actually I prefer to leave `adjoin_simple_le_iff` as it is used more in this file latter, is it OK?